ion-schema/src/model/ranges.rs

```diff
@@ -0,0 +1,345 @@
+            let Some(seq) = ion.as_list() else {
+                invalid_schema_error(format!("range must be a non-null list; found: {ion}"))?
+            };
+
+            isl_require!(seq.len() == 2 => "range must have a lower and upper bound; found: {ion}")?;
+            // We can unwrap these without panicking
+            let lower = seq.get(0).unwrap();
+            let upper = seq.get(1).unwrap();
+
+            let lower = match (lower.as_symbol_text(), lower.one_optional_annotation()?) {
+                (Some(MIN), Some(EXCLUSIVE)) => invalid_schema_error("'min' may not be exclusive")?,
+                (Some(MIN), None) => Unbounded,
+                (_, Some(EXCLUSIVE)) => Excluded(T::try_read(lower, ctx)?),
+                (_, None) => Included(T::try_read(lower, ctx)?),
+                _ => invalid_schema_error(format!("invalid annotation on range bound: {lower}"))?,
+            };
+            let upper = match (upper.as_symbol_text(), upper.one_optional_annotation()?) {
+                (Some(MAX), Some(EXCLUSIVE)) => invalid_schema_error("'max' may not be exclusive")?,
+                (Some(MAX), None) => Unbounded,
+                (_, Some(EXCLUSIVE)) => Excluded(T::try_read(upper, ctx)?),
+                (_, None) => Included(T::try_read(upper, ctx)?),
+                _ => invalid_schema_error(format!("invalid annotation on range bound: {upper}"))?,
+            };
+
+            Ok(IonSchemaRange::try_new(lower, upper)?)
+        } else {
+            isl_require!(optional_annotation.is_none() => "invalid annotation on range; found: {ion}")?;
+            let value: T = T::try_read(ion, ctx)
+                .map_err(|_| invalid_schema_error_raw(format!("invalid range value: {ion}")))?;
+            Ok(IonSchemaRange::from(value))
+        }
+    }
+}
+
+impl<T: RangeBoundType> From<&T> for IonSchemaRange<T> {
+    fn from(value: &T) -> Self {
+        IonSchemaRange {
+            min: Included(value.clone()),
+            max: Included(value.clone()),
+        }
+    }
+}
+impl<T: RangeBoundType> From<T> for IonSchemaRange<T> {
+    fn from(value: T) -> Self {
+        IonSchemaRange {
+            min: Included(value.clone()),
+            max: Included(value),
+        }
+    }
+}
+
+// TODO: Once https://github.com/amazon-ion/ion-schema/issues/140 is resolved, revisit this and
+//       possibly replace with a `TryFrom` implementation.
+macro_rules! from_range {
+    ($range_type:ident) => {
+        impl<T: RangeBoundType> From<$range_type<T>> for IonSchemaRange<T> {
+            fn from(value: $range_type<T>) -> Self {
+                let min = value.start_bound().cloned();
+                let max = value.end_bound().cloned();
+                IonSchemaRange { min, max }
+            }
+        }
+    };
+}
+from_range!(Range);
+from_range!(RangeTo);
+from_range!(RangeFrom);
+from_range!(RangeInclusive);
+from_range!(RangeToInclusive);
+
+#[cfg(test)]
+mod tests {
```

Comment:
I couldn't fin examples of ranges with timestamp precision in the timestamp_precision constraint file. Maybe we should add those different ranges here to check for read and write.(probably something like range::[month, year]).

Reply:
Sure, I can add some.